casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala

```diff
@@ -32,6 +33,53 @@ object DagOperations {
     StreamT.delay(Eval.now(build(Queue.empty[A].enqueue[A](start), HashSet.empty[A])))
   }
 
+  /**
+    * Determines the ancestors to a set of blocks which are not common to all
+    * blocks in the set. Each starting block is assigned an index (hence the
+    * usage of IndexedSeq) and this is used to refer to that block in the result.
+    * A block B is an ancestor of a starting block with index i if the BitSet for
+    * B contains i.
+    * @param blocks indexed sequence of blocks to determine uncommon ancestors of
+    * @param lookup association of block hashes to data about the block (for DAG
+    *               traversal)
+    * @param topoSort topological sort of the DAG, ensures ancestor computation is
+    *                 done correctly
+    * @return A map from uncommon ancestor blocks to BitSets, where a block B is
+    *         and ancestor of starting block with index i if B's BitSet contains i.
+    */
+  def uncommonAncestors(blocks: IndexedSeq[BlockMetadata], lookup: BlockMetadata.Lookup)(
+      implicit topoSort: Ordering[BlockMetadata]): Map[BlockMetadata, BitSet] = {
+    def parents(b: BlockMetadata): List[BlockMetadata] = b.parents.map(lookup)
+    def isCommon(set: BitSet): Boolean                 = (0 until blocks.length).forall(set.contains)
+
+    val initMap = blocks.zipWithIndex.map { case (b, i) => b -> BitSet(i) }.toMap
+    val q       = new mutable.PriorityQueue[BlockMetadata]()
+    q.enqueue(blocks: _*)
+
+    @tailrec
+    def loop(currMap: Map[BlockMetadata, BitSet],
+             visited: HashSet[BlockMetadata]): Map[BlockMetadata, BitSet] =
+      if (q.isEmpty) currMap
+      else {
+        val currBlock = q.dequeue()
+        val currSet   = currMap.getOrElse(currBlock, BitSet.empty)
+        if (visited(currBlock) || isCommon(currSet)) {
```

Comment:
If it's common, you can remove it from the map. (You won't see it again because you're going in reverse topological order)

Reply:
Also you should only need to test visited on insert, not on removal.